[Multi-AZ] Modify networking validators impacted by Multi-AZ

jdeamicis opened the pull request:
### Description of changes
* Modify Networking validators impacted by the Multi-AZ functionality in the Slurm and Scheduler Plugin scenarios:
  * Add better description for `SubnetsValidator`.
  * Add new `QueueSubnetsValidator` to check
    * if there are duplicate IDs in `subnet_ids`;
    * if all subnets in a queue are from different AZs;
  * Modify `SingleSubnetValidator` to:
    * remove check on unique subnet across all compute queues;   
    * check if compute resources with single instance type have one single subnet;
  * Make the implementation of `compute_subnet_ids` more readable

### Tests
* Added or modified unit tests for the validator impacted by this PR
* No change was introduced in the core functioning of the product, so except for the validation phase the behavior of the product is the same.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] ~Check if documentation is impacted by this change.~

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Add better description to SubnetsValidator

Add comment about usage of SubnetsValidator in the cluster config validation.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add QueueSubnetsValidator

The new validator checks that the subnets defined in a queue are all in
different AZs because the EC2 Fleet does not support using multiple
subnets from the same availability zone.

The new validator applies to Slurm and Scheduling Plugin queues.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Check for duplicate entries in a queue's SubnetIds

Fail cluster creation if the SubnetIds parameter contains duplicate
subnet IDs.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Check for multiple subnets in single instance type scenarios

Modify SingleSubnetValidator to be used to check that only 1 subnet is
provided in queues containing compute resources with single instance type.

Remove check for single subnet across all queue for Slurm and Scheduler
Plugin scenarios.

Modify the unit tests for the validator accordingly.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Make compute_subnet_ids more readable

The previous implementation was pythonic but a bit hard to read.
Change to an equivalent, more readable implementation (speed is
not fundamental here).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Small improvement to network validators impacted by Multi-AZ

Changed some validator messages and renamed the test function to test
QueueSubnetsValidator.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

EddyMM commented on cli/src/pcluster/validators/networking_validators.py:
```diff
@@ -57,6 +63,43 @@ def _validate(self, subnet_ids: List[str]):
+
+        # Test if the subnets are all in different AZs
+        try:
+            subnets = AWSApi.instance().ec2.describe_subnets(subnet_ids=subnet_ids)
```
This logic of getting the Subnet AZs can moved into the `ec2.py` file for re-use or extend the existing implementation [get_subnet_avail_zone](https://github.com/EddyMM/aws-parallelcluster/blob/50264bbdbf1c540dbe916db7e007b8c7c0e7b6f1/cli/src/pcluster/aws/ec2.py#L113-L113).

I've seen a similar case in the EFA Validator: [az_list](https://github.com/aws/aws-parallelcluster/pull/4544/files#diff-bd4426f9ee10414432866ef60f1b6a41dab292da908dc94ded28ffee6cc66cb3R1912)

EddyMM replied:
I think it would be a better experience if the customer was informed about which subnets are in the same AZ, especially since we have this information within this code block.

For duplicate subnets, it's easy for the user to see the duplicate (since there are few AZs/Subnets that can be included).

jdeamicis replied:
> This logic of getting the Subnet AZs can moved into the `ec2.py` file for re-use or extend the existing implementation [get_subnet_avail_zone](https://github.com/EddyMM/aws-parallelcluster/blob/50264bbdbf1c540dbe916db7e007b8c7c0e7b6f1/cli/src/pcluster/aws/ec2.py#L113-L113).
> 
> I've seen a similar case in the EFA Validator: [az_list](https://github.com/aws/aws-parallelcluster/pull/4544/files#diff-bd4426f9ee10414432866ef60f1b6a41dab292da908dc94ded28ffee6cc66cb3R1912)

This has been addressed in #4547, on top of which this branch has been rebased.

jdeamicis replied:
> I think it would be a better experience if the customer was informed about which subnets are in the same AZ, especially since we have this information within this code block.
> 
> For duplicate subnets, it's easy for the user to see the duplicate (since there are few AZs/Subnets that can be included).

I added both lists, for more clarity.

jdeamicis committed:
Use subnet_id_az_mapping in QueueSubnetsValidator

Use the new mapping property to get the AZs of the subnets in a queue.
Adapt the unit test accordingly.

Small modification to the validation message for QueueSubnetsValidator.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

EddyMM commented on cli/src/pcluster/validators/networking_validators.py:
```diff
@@ -57,6 +63,39 @@ def _validate(self, subnet_ids: List[str]):
+
+            if len(az_set) < len(subnet_ids):
+                self._add_failure(
+                    "SubnetIds specified in queue {0} contains multiple subnets in the same AZ. "
```
From previous revision

> I think it would be a better experience if the customer was informed about which subnets are in the same AZ, especially since we have this information within this code block.

It could be cumbersome for a customer to check each subnet one by one again to find out which ones are in the same AZ.

jdeamicis replied:
I addressed this (see [above](https://github.com/aws/aws-parallelcluster/pull/4541#discussion_r1025403078))

EddyMM commented on cli/src/pcluster/validators/networking_validators.py:
```diff
@@ -69,12 +108,17 @@ def _validate(self, elastic_ip: Union[str, bool]):
 
-    def _validate(self, queues_subnets):
-        subnet_ids = {tuple(set(queue_subnets)) for queue_subnets in queues_subnets}
+    def _validate(self, queue_name, subnet_ids):
```
A validator should be able to determine the correct/incorrect state using only the information it is provided.
This validator already assumes that the queue is being used with a Single Instance Type (since it's only registered when there is a Single Instance Type usage).

I think validation logic should be local to the validator. The list of ComputeResources can be passed to this validator and it should then determine what's valid/invalid.

EddyMM replied:
This will also allow the validator to communicate which Compute Resource is invalid.

jdeamicis replied:
I had to do this to avoid a circular import between `cluster_config.py` and `networking_validators.py`. We can discuss about this.

jdeamicis committed:
Address Code Review comments

Print duplicate subnet IDs in QueueSubnetsValidator.

Print AZs that have multiple subnets used in a queue in
QueueSubnetValidator.

Modify the validator test function accordingly.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

EddyMM commented on cli/src/pcluster/validators/networking_validators.py:
```diff
@@ -57,6 +64,58 @@ def _validate(self, subnet_ids: List[str]):
+                if len(az_set) < len(subnet_ids):
+
+                    # Find the AZs with multiple subnets
+                    azs_with_multiple_subnets = {}
```
Could we move this logic into a separate function, named `get_subnets_sharing_an_az` perhaps.

Also, could it be simplified to avoid the nested loop? Would this logic work for this use-case?
```
def get_subnets_sharing_an_az(subnet_ids):
    subnets_sharing_az = set()
    visited_subnet = {}

    for subnet_id in subnet_ids:
        az = subnet_id_az_mapping[subnet_id]

        if visited_subnet.get(subnet_id) == az:
            subnets_sharing_az.add(subnet_id)
        else:
            visited_subnet[subnet_id] = az

    return list(subnets_sharing_az)
```

jdeamicis replied:
We discussed about this offline and decided to implement a new property `az_subnet_ids_mapping` in the `_QueueNetworking` class. This property is a dictionary containing the list of subnets belonging to each AZ used in the queue.

jdeamicis committed:
Add az_subnet_ids_mapping property to _QueueNetworking class

Use this property to quickly check if any AZ is used by multiple subnets
in QueueSubnetsValidator, simplifying the implementation.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cli/src/pcluster/validators/networking_validators.py:
```diff
@@ -57,6 +64,55 @@ def _validate(self, subnet_ids: List[str]):
+                FailureLevel.ERROR,
+            )
+
+        # Test if the subnets are all in different AZs
```
Are we sure these tests are mutually exclusive ?

It seems possible to have duplicated ID in the list of subnets and also have more than one subnet-id in the same AZ.

jdeamicis replied:
No, they are not mutually exclusive. I just don't want the duplicate subnet IDs to trigger also the second part of the validator. Without this `if ... else` the duplicate subnet IDs would also be counted as multiple subnets in the same AZ, which to me doesn't make sense.

jdeamicis replied:
But I see what you mean. I would need to remove the duplicate subnet IDs and pass this intermediate list of subnet IDs to the next phase of the validator.

Is this worth the effort? The validation would fail anyway.

jdeamicis replied:
Good catch @NSsirena. I changed implementation so that now it's possible to have both validation messages at the same time.

jdeamicis committed:
Implement az_subnet_ids_mapping as dictionary of sets

This allows to make the two parts of QueueSubnetsValidator
completely orthogonal to each other. Now an error about
multiple subnets defined in the same AZ would be printed even
if there are duplicate subnet ids in the QueueNetworking definition.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
